### ISSUE TITLE
Add support for dynamically resizing volumes provisioned with rook

### ISSUE THREAD

**Issue by carldanley** [rook/rook#1169]
Kubernetes 1.8 added the ability to resize persistent volume claims (https://kubernetes.io/docs/admin/admission-controllers/#persistentvolumeclaimresize). I think it would be really nice to have support for dynamically growing (and maybe shrinking, if possible) volumes provisioned with rook.

**Comment by kokhang** [comment]
More info here: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims

**Comment by DanKerns** [comment]
+1. I think this is a super natural thing for the Rook Operator:  The underlying mechanism is complex and needs to be carefully orchestrated.

**Comment by jbw976** [comment]
@kokhang so with the `ExpandPersistentVolumes` feature gate set, when a user edits a PVC to update the size, does the flexvolume plugin get called? what part of the flexvolume interface would handle that? I'm not sure how this all works under the covers.

**Comment by kokhang** [comment]
@jbw976 the flexvolume is not in the path of resizing. It would be something handled by the provisioner. We might have to add this support to the external-provisioner library.

**Comment by edillmann** [comment]
any progress on this ?

**Comment by edillmann** [comment]
FYI I found this reference on kubernetes side : https://github.com/kubernetes/community/pull/1700

**Comment by jbw976** [comment]
This could be possible to implement in the volume provisioner, and may already have some logic in the upstream provisioner controller in Kubernetes external storage repo: https://github.com/kubernetes-incubator/external-storage/blob/master/lib/controller/controller.go

According to the sig-storage planning sheet, volume expansion will go to beta in k8s 1.11: https://docs.google.com/spreadsheets/d/1t4z5DYKjX2ZDlkTpCnp18icRAQqOE85C1T1r2gqJVck/edit#gid=1308589045

We should also remember that volume resizing looks like it will be included in CSI also, so that could be a path for implementing this instead.

It'd be nice if someone from the community wanted to jump in and take on this feature :)

**Comment by aleks-mariusz** [comment]
@voron per @kokhang flexvolume is not in the path of resizing, so is there something else holding this particular feature of rook-ceph performing resizes up ?

**Comment by aleks-mariusz** [comment]
in any case, it looks like it made it into v1.13 per https://github.com/kubernetes/enhancements/issues/304#issuecomment-438019795 so can any progress be made to implementing this in ceph-rook now please? :-)

**Comment by MatthiasLohr** [comment]
Yeah, please! :)

**Comment by imdingtalk** [comment]
+1   i need

**Comment by travisn** [comment]
@rootfs does the csi driver take care of resizing?

**Comment by gabibbo97** [comment]
> @rootfs does the csi driver take care of resizing?

https://github.com/ceph/ceph-csi/issues/101

**Comment by luxcn** [comment]
Yeah, i need too.

**Comment by Mengkzhaoyun** [comment]
+1 i need

**Comment by RiaanLab** [comment]
+1 For resizing in Rook. The rest of Rook is GREAT, though :-)

**Comment by rootfs** [comment]
For CSI driver, there was a dependency to wait for external-resizer. Now that dependency is resolved. once ceph-csi supports volume expansion, Rook should be able to pick it up. 

cc @travisn

**Comment by BlaineEXE** [comment]
Ceph CSI will be beta in Rook v1.0 release with the goal of stable in v1.1. We are therefore expecting that as long as the CSI beta goes well, this feature will be stable in v1.1, though it may not be available in the v1.0 release.

**Comment by jonashartwig** [comment]
Hi, I see that dynamic resizing is not available yet. My PVC is full and i would like to replace the PVC of some pod with a new one and keeping the content. Can you give me some ideas on how to do this please? regards!!!

**Comment by dimm0** [comment]
https://github.com/rook/rook/issues/3133#issuecomment-491941697

**Comment by jonashartwig** [comment]
amazing, thank you!

**Comment by davi5e** [comment]
> #3133 (comment)

Broken link, not working anymore...

Also, it would appear [ceph: Add dynamic flexvolume resize · Issue #3417](https://github.com/rook/rook/issues/3417) is also broken and could hep this issue...

**Comment by aleks-mariusz** [comment]
> > #3133 (comment)
> 
> Broken link, not working anymore...
> 
> Also, it would appear [ceph: Add dynamic flexvolume resize · Issue #3417](https://github.com/rook/rook/issues/3417) is also broken and could hep this issue...
> 
> Are these links really broken or is it some misconfig in my side?

neither.. [issues with github](https://www.githubstatus.com/incidents/tq7pm7mn9b1k) are the culrit

**Comment by shuaiwang8765** [comment]
> For CSI driver, there was a dependency to wait for external-resizer and kubernetes csi volume to support expand. Now the first dependency is resolved. Once ceph-csi supports volume expansion, Rook should be able to pick it up.
> 
> cc @travisn
hello, ceph offical has supported csi-resizer,so I want to know wheather  the rook can support csi volume expand .  what the time?

**Comment by BlaineEXE** [comment]
@Madhu-1 do you have an official answer here?

**Comment by Madhu-1** [comment]
Cephfs and rbd resize will be part of cephcsi v2.0.0 release. Refer https://github.com/ceph/ceph-csi/issues/557 for more info

**Comment by BlaineEXE** [comment]
It seems this is not complete with CSI then, so reopening.

**Comment by BlaineEXE** [comment]
It seems this is not complete with CSI then since it looks like Rook currently uses ceph-csi v1.2.2, so reopening.

**Comment by Madhu-1** [comment]
Assigning to myself, once cephcsi 2.0.0 release is done, will make required changes in rook to support resize

**Comment by paalkr** [comment]
cephcsi 2.0.0 was releases yesterday :)
https://github.com/ceph/ceph-csi/releases/tag/v2.0.0

**Comment by Madhu-1** [comment]
am working on a PR to update ceph-csi in rook

**Comment by onedr0p** [comment]
@BlaineEXE are we awaiting a release to the rook helm chart inorder to start using this?

**Comment by travisn** [comment]
@onedr0p The plan is to backport this to the release-1.2 branch and include the CSI 2.0 driver in the next patch release, perhaps by end of week. It won't be enabled by default yet, but will have instructions on how to enable it in v1.2 clusters. See [ceph-upgrade.md](https://github.com/rook/rook/blob/master/Documentation/ceph-upgrade.md#enabling-the-csi-v20-driver) for the instructions. In the next rook release (v1.3), it will be enabled by default.
